Speed improvements

FaBay89 opened the pull request:
I dont see any needs for the following functions, maybe you know why they should not be removed?
I tested only in dev enviroment (python 3.4.3, django 1.9), but could not see any changes in memory usage. 

- reset_queries() and connection.close() each saves ~100s.
- progress.update() does not work for me anyway, i dont know why, so it saves ~100s.
- _travis(), ok, but also saves ~100s.

All in all i reduced the runtime from about 700s to 300s. Quite a lot.

jpic commented:
IIRC reset_queries was for ram useage.

_travis is not needed anymore, now that we test imports with django-dbdiff.

we might want to leave progress update an option thought.

FaBay89 commented:
So you want reset_queries to stay?

Does progressbar works for you? I just get the finished bar after the process, but there is nothing to see while processing... I will try to call update only every (max/1000)th iteration, maybe this saves enough time. And 0.1% should be enough accuracy.

What about connection.close()?

max-arnold commented:
@FaBay89 Regarding update() speed, could you please try older revision before PR #99 and #105 were merged? I was sure I improved the speed by sending this PR to progressbar2 library https://github.com/WoLpH/python-progressbar/pull/43.

Connection.close() seems to be MySQL related: e7f69af01760c450b4a72db84fda3d98d6731928

reset_queries() is probably helpful only for `DEBUG=True` mode, so it could be made conditional:

```
if settings.DEBUG:
    reset_queries()
```

FaBay89 commented:
progressbar.update(): Ok, but before I try to get progressbar running.. Maybe this is the time issue already. The examples of progressbar2 are also not showing up while processing.

connection.close(): From django ticket:
> RECOMMENDED SOLUTION: close the connection with from django.db import connection; 
connection.close() when you know that your program is going to be idle for a long time. 

I still see no reason to close the connection on every parsed line (about 10M times)?

max-arnold commented:
Yep, connection.close() could be called less often.

jpic commented:
Yeah I don't remember how this got there, :+1: for removing it. Nice work BTW ;0

FaBay89 commented:
Ok, i removed `_travis()`, `connection.close()` and unused `import` statements.

`reset_queries()` is conditional now by django DEBUG setting, or did you mean a explicit DEBUG setting in cities_light @max-arnold?

I tested the progressbar again (it's working now) and it seems that it runs into issues by to many `update()` calls. I fixed the behaviour to call it only on every changed 0.1%.
This improved runtime from >80sec to <1sec for the update calls.

max-arnold commented:
@FaBay89 You are right - it is the settings.DEBUG, see the docs https://docs.djangoproject.com/en/1.9/faq/models/#how-can-i-see-the-raw-sql-queries-django-is-running (there is nothing to reset when settings.DEBUG is False, so there is no memory leak).

Regarding the progressbar2, have you looked at _needs_update() method? https://github.com/WoLpH/python-progressbar/blob/develop/progressbar/bar.py#L407  I'm really interested why it does't work for you. It should perform approximately as many updates as there are horizontal characters in your terminal, so there is no point to divide the value by 1000. When you call `update()` in most cases the code path should be quite short, from line https://github.com/WoLpH/python-progressbar/blob/develop/progressbar/bar.py#L429 to line https://github.com/WoLpH/python-progressbar/blob/develop/progressbar/bar.py#L451. The overhead of this short path can be still noticeable if you call `update()` a lot, but then it is much better to disable progressbar completely (`--disable-progressbar` or something like this).

I did a benchmark and on my machine clean import without `update_interval` patch took 22:22, with `update_interval` patch it took 20:06 (I haven't disabled `reset_queries()` and `_travis()` to test the effect in isolation).

Also with `update_interval` patch I see no progressbar updates at all (it just shows 100% at the end of operation), so it basically disabled it completely. Tested with `progressbar2==3.6.0`.

max-arnold commented:
Before the patch:

```
i += 1
progress.update(i)
```

After the patch:

```
if i == 0 % update_interval:
    progress.update(i)
i += 1
```

`0 % update_interval` is always zero, so `i == 0` only once (since you moved `i += 1` after the `update()` call).

FaBay89 commented:
At first: added a new improvement. In `translation_parse()` i changed the type of `country_ids`, `region_ids` and `city_ids` from `list()` to `set()`, because searching in sets is blazingly faster than in lists.

I installed progressbar2 version 3.6.1 from git, which solves the display issue.

Ok big mistake at my side, of course it should be
````
if 0 == i % update_interval:
      progress.update(i)
i += 1
````

I am still looking for the overhead...

FaBay89 commented:
Overhead:
`update` is called 
~12M times; Time: 77s, Own-Time: 25s
`_needs_update`:
~12M times; Time: 52s, Own-Time: 32s
`<built-in method now>` (`Datetime.now()`):
~12M times; Time: 20s, Own-Time: 20s

So the `25s + 32s + 20s = 77s` are the runtime of the update method (77s are ~18% of complete runtime from cities_light in my case) and is generated only by calculations in these 3 methods.

If you are not happy with the 0.1% we could call `update()` on every 0.01% or 0.001%.
With 0.001% it still will be called ~1.2M times (estimated), but needs only ~7s (estimated). But i would prefer a higher percentage, cause you never need so much accuracy in process display.

There is no difference you can see, but it is much faster, just by do not call update so many times, so why you not happy with this?

max-arnold commented:
What do you think about disabling the progressbar completely with an optional command line switch?

max-arnold commented:
For example `--no-progress` or `--progress` (if disabled by default) option?  @jpic

FaBay89 commented:
I think that this is not necessary, because it is a nice feature and can be used without overhead... There is no disadvantage by using the patch, or why do you think it is better to disable it?

jpic commented:
I think progress can be disabled by default.

I don't **want** reset_queries, maybe it's not needed anymore.

Don't worry so much about it and do your best, just make sure you're ready to setup mysql and test it when people report issues with it again :D

FaBay89 commented:
But disabling progressbar do not solve the overhead when you enable it, so why you do not want to fix the overhead?

jpic commented:
I didn't say I don't *want* to fix anything, go for it ;)

FaBay89 commented:
Is done already in my commit ;)

What about `reset_queries`? Remove it

jpic commented:
Remove it and let's see if anyone complains about it ie. debugging or mysql.

jpic commented:
LGTM, anybody wants to go ahead and deploy this ?

jpic commented:
@max-arnold should we have this ?

max-arnold commented:
@jpic I'm +0 on this one.

max-arnold commented:
I refactored this PR into separate branch https://github.com/max-arnold/django-cities-light/commits/feature/speedups. There are some some differences though:

* reset_queries is here, but used only in DEBUG mode
* connection.close() is called only for MySQL
* worst case with progressbar is not optimized (left as is), but it is disabled by default

My angle is that I do not want extreme optimizations which increase code complexity (there are already too many IF statements and the command is quite monolithic) and lead to regressions (memory leaks, broken connections on MySQL). At the same time, best case should be quite close to what @FaBay89 wants. Also I do not care a lot about initial import speed because we now have cities_light_fixtures command for fast and repeatable imports.

Below is my completely unscientific benchmark (ran only once, huge variation of parallel disk activity etc...):

```
stable:                           1179,28s user 28,34s system 92% cpu 21:49,57 total
removed travis:                   1167,87s user 25,26s system 91% cpu 21:46,49 total
disabled debug and reset_queries: 1123,67s user 21,87s system 92% cpu 20:39,04 total
disabled progressbar:             993,14s  user 20,95s system 90% cpu 18:43,77 total
disabled connection.close:        961,46s  user 21,34s system 91% cpu 17:48,74 total
lists to sets:                    507,23s  user 11,64s system 86% cpu 10:03,06 total
```

I plan to refactor cities_light command in the future and discussed possible approaches with James on EuroPython 2016 so it may be possible to speedup it further. As for now, I'd like to minimize extreme optimizations which can complicate planned refactoring.

@FaBay89, @jpic Any feedback on my branch?

jpic commented:
Looks very good !

max-arnold commented:
Ok, I'll merge it in a few days if I get no feedback from original author.

max-arnold commented:
Fixed in #131
